### PR TITLE
lestarch: adding in new ground framing components

### PR DESCRIPTION
**Pull request by LeStarch**
Co-authored-by: M Starch <LeStarch@googlemail.com>
Co-authored-by: saba-ja <saba.janamian.663@my.csun.edu>

| | |
|:---|:---|
|**_Originating Project/Creator_**| Infrastructure |
|**_Affected Component_**| Svc |
|**_Affected Architectures(s)_**|  |
|**_Related Issue(s)_**|  |
|**_Has Unit Tests (y/n)_**| y |
|**_Builds Without Errors (y/n)_**| y  |
|**_Unit Tests Pass (y/n)_**| y |
|**_Documentation Included (y/n)_**| y |

---
## Change Description

Adds in refactored and more flexible framing components.

## Rationale

The old "GroundInterface" was not reusable nor flexible.

### CODE REVIEW

File: Svc/Deframer/test/ut/Tester.cpp [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+// Copyright 2009-2021, by the California Institute of Technology.
+// ALL RIGHTS RESERVED.  United States Government Sponsorship
+// acknowledged.
+// 
+// ======================================================================
+
+#include "Tester.hpp"
+#include <iostream>
+#include <stdio.h>
+using namespace std;
```
Comment (saba-ja) [inline Svc/Deframer/test/ut/Tester.cpp:16]: There is no need for these. They can be safely removed.

File: Svc/FramingProtocol/docs/sdd.md [hunk 2]
```diff
@@ -0,0 +1,40 @@
+\page SvcFrameringProtocol FramerProtocol
+# Svc::FrameringProtocol FrameringProtocol Component
```
Comment (kevin-f-ortega) [inline Svc/FramingProtocol/docs/sdd.md:2]: First line and second line: change FrameringProtocol to FramingProtocol.

File: Svc/Deframer/docs/sdd.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
+\page SvcDeframer Deframer Component
+# Svc::Deframer Deframer Component
+
+Deframer takes a stream of bytes from ground and after checking the header and footer extracts data from the stream. The extracted data usually is passed to the service layer (typically CmdDisp, FileUplink, or a GenericHub). The Deframer uses a `DeframingProtocol` for the same abstraction that the Framer uses `FramingProtocol`. Deframer is backed by a circular buffer such that incomplete messages are reassembled. It also can be hooked up in two ways:
```
Comment (kevin-f-ortega) [inline Svc/Deframer/docs/sdd.md:4]: suggestions: add comma after `ground and` and after `the header` in the first sentence.

**Review by kevin-f-ortega — COMMENTED** [review]
Overall PR looks good to me. Just a few comments.

**Comment by LeStarch** [comment]
Now depends on #399 to properly handle time-get.

File: Svc/FramingProtocol/FprimeProtocol.cpp [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+const FP_FRAME_TOKEN_TYPE FprimeFraming::START_WORD = static_cast<FP_FRAME_TOKEN_TYPE>(0xdeadbeef);
+
+FprimeFraming::FprimeFraming(): FramingProtocol() {}
+
+FprimeDeframing::FprimeDeframing(): DeframingProtocol() {}
+
+void FprimeFraming::frame(const U8* const data, const U32 size, Fw::ComPacket::ComPacketType packet_type) {
+    FW_ASSERT(m_interface != NULL);
+    FP_FRAME_TOKEN_TYPE real_data_size = size + ((packet_type != Fw::ComPacket::FW_PACKET_UNKNOWN) ? sizeof(I32) : 0);
```
[the remark this answers is not in the document]
Comment (LeStarch) [reply Svc/FramingProtocol/FprimeProtocol.cpp:18]: Here we need to use I32 specifically.

**Comment by LeStarch** [comment]
@kevin-f-ortega  and @saba-ja I think these issues are now resolved at your convenience, would you take a another look?

**Review by kevin-f-ortega — APPROVED** [review]
LGTM.

**Pull request merged**